### ISSUE TITLE
BulkInsertOrUpdate Version 7.1.6 throw error  "CREATE INDEX CONCURRENTLY cannot run inside a transaction block" 

### ISSUE THREAD

**Issue by yehiaelsayed** [borisdj/EFCore.BulkExtensions#1276]
**Problem:** 
I was using the function BulkInsertOrUpdate with `UpdateByProperties`, and it worked fine on version `7.0.1 ` when I updated the package to version 7.1.6 it started throwing this error **CREATE INDEX CONCURRENTLY cannot run inside a transaction block**
the table has no key so there is no need to exclude the identity column and the table **has a unique index** on the `UpdateByProperties `

**Update:** 
This issue started with version 7.1.1 

**Other Packages Versions** 

"Microsoft.EntityFrameworkCore" Version="7.0.11" 
"Npgsql.EntityFrameworkCore.PostgreSQL" Version="7.0.11" 

Function call example 
```
 dbContext.BulkInsertOrUpdate(date, new BulkConfig()
 {
     UpdateByProperties = new List<string>() { "Col1", "Col2", "Col3" },

 });
```
Exception StackTrace 

```

Npgsql.PostgresException (0x80004005): 25001: CREATE INDEX CONCURRENTLY cannot run inside a transaction block
   at Npgsql.Internal.NpgsqlConnector.<ReadMessage>g__ReadMessageLong|233_0(NpgsqlConnector connector, Boolean async, DataRowLoadingMode dataRowLoadingMode, Boolean readingNotifications, Boolean isReadingPrependedMessage)
   at Npgsql.NpgsqlDataReader.NextResult(Boolean async, Boolean isConsuming, CancellationToken cancellationToken)
   at Npgsql.NpgsqlDataReader.NextResult(Boolean async, Boolean isConsuming, CancellationToken cancellationToken)
   at Npgsql.NpgsqlDataReader.NextResult()
   at Npgsql.NpgsqlCommand.ExecuteReader(CommandBehavior behavior, Boolean async, CancellationToken cancellationToken)
   at Npgsql.NpgsqlCommand.ExecuteReader(CommandBehavior behavior, Boolean async, CancellationToken cancellationToken)
   at Npgsql.NpgsqlCommand.ExecuteNonQuery(Boolean async, CancellationToken cancellationToken)
   at Npgsql.NpgsqlCommand.ExecuteNonQuery()
   at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteNonQuery(RelationalCommandParameterObject parameterObject)
   at Microsoft.EntityFrameworkCore.RelationalDatabaseFacadeExtensions.ExecuteSqlRaw(DatabaseFacade databaseFacade, String sql, IEnumerable`1 parameters)
   at Microsoft.EntityFrameworkCore.RelationalDatabaseFacadeExtensions.ExecuteSqlRaw(DatabaseFacade databaseFacade, String sql, Object[] parameters)
   at EFCore.BulkExtensions.SqlAdapters.PostgreSql.PostgreSqlAdapter.MergeAsync[T](DbContext context, Type type, IEnumerable`1 entities, TableInfo tableInfo, OperationType operationType, Action`1 progress, Boolean isAsync, CancellationToken cancellationToken)
   at EFCore.BulkExtensions.SqlAdapters.PostgreSql.PostgreSqlAdapter.MergeAsync[T](DbContext context, Type type, IEnumerable`1 entities, TableInfo tableInfo, OperationType operationType, Action`1 progress, Boolean isAsync, CancellationToken cancellationToken)
   at EFCore.BulkExtensions.SqlAdapters.PostgreSql.PostgreSqlAdapter.Merge[T](DbContext context, Type type, IEnumerable`1 entities, TableInfo tableInfo, OperationType operationType, Action`1 progress)
   at EFCore.BulkExtensions.SqlBulkOperation.Merge[T](DbContext context, Type type, IEnumerable`1 entities, TableInfo tableInfo, OperationType operationType, Action`1 progress)
   at EFCore.BulkExtensions.DbContextBulkTransaction.Execute[T](DbContext context, Type type, IEnumerable`1 entities, OperationType operationType, BulkConfig bulkConfig, Action`1 progress)
   at EFCore.BulkExtensions.DbContextBulkExtensions.BulkInsertOrUpdate[T](DbContext context, IEnumerable`1 entities, BulkConfig bulkConfig, Action`1 progress, Type type)
  Exception data:
    Severity: ERROR
    SqlState: 25001
    MessageText: CREATE INDEX CONCURRENTLY cannot run inside a transaction block
    File: xact.c
    Line: 3413
    Routine: PreventInTransactionBlock

```

**Comment by yehiaelsayed** [comment]
@borisdj 
I was using Version 7.0.1 and everything was working fine then I had to update to the latest version because Geometry Types was not supported in version 7.0.1  and then the issue **CREATE INDEX CONCURRENTLY cannot run inside a transaction block**
started to happen because the library for some reason is not able to detect the unique constrain on the table and try to create a unique constrain on the fly but because the insert statement is within the transaction it throws the previous error I tried to keep downgrading in the version and I found it started to happen to start from version 7.1.1 
**Model:**
```
 public class DeviceLocation 
 {
     public int Id { get; set; }
     public int DeviceId { get; set; }
     public Guid ProjectId { get; set; }
     public DateTime GeneratedAt { get; set; }
 }
```
**Model Configuration:**
```
public class DeviceLocationConfiguration : IEntityTypeConfiguration<DeviceLocation>
{
    public void Configure(EntityTypeBuilder<DeviceLocation> builder)
    {
       builder.HasNoKey();
       
       builder.HasIndex(dl => new { dl.ProjectId, dl.DeviceId, dl.GeneratedAt }).IsUnique(true);

       builder.Property(e => e.Id).UseIdentityAlwaysColumn();
       builder.Property(e=>e.Id).ValueGeneratedOnAdd();	 
       builder.Property(dl => dl.DeviceId);
       builder.Property(dl => dl.ProjectId);
       builder.Property(dl => dl.GeneratedAt).HasColumnType(NpgsqlDbType.Timestamp.ToString().ToLower());
       builder.HasIndex(dl => new { dl.DeviceId, });
       builder.HasIndex(dl => new { dl.ProjectId, dl.GeneratedAt });
       builder.HasIndex(dl => dl.GeneratedAt);
     

    }
}
```
**Insert Data:**
```
  var pgConnStr ="my connection string"
  var pgOptionsBuilder = new DbContextOptionsBuilder<MyDbContext>();

  pgOptionsBuilder.UseNpgsql(pgConnStr, o => o.UseNetTopologySuite());
  var dbContext = new MyDbContext(pgOptionsBuilder.Options);
var transaction =  dbContext.Database.BeginTransaction();

  var deviceLocationDate  = new List<DeviceLocation>(); 
deviceLocationDate.Add(new  DeviceLocation(){ DeviceId=1 , ProjectId=Guid.Parse("60E97E07-E528-4DA5-A7C3-7F798405FB9B"),GeneratedAt=DateTime.Now  } );

  dbContext .BulkInsertOrUpdate(deviceLocationDate, new BulkConfig()
  {
      PropertiesToExclude = new() { "Id", "UpdatedAt" },
      UpdateByProperties = new() { "ProjectId","DeviceId", "GeneratedAt", },
  });
  
  transaction.Commit();
```

**Comment by xzxzxc** [comment]
It seems like I found the issue - in the `PostgreSqlQueryBuilder.CountUniqueConstrain`, there is a check only for constraints, but an index could also be marked as unique, producing identical uniqueness behaviour.

So, as a workaround, you could add a new dummy constraint
```c#
/// <inheritdoc />
public partial class AddDummyUniqueConstraintWorkaround : Migration
{
	// This is a workaround for https://github.com/borisdj/EFCore.BulkExtensions/issues/1276

	/// <inheritdoc />
	protected override void Up(MigrationBuilder migrationBuilder)
	{
		migrationBuilder.Sql("""
ALTER TABLE "table_name"
    ADD CONSTRAINT constraint_name
        UNIQUE ("col1", "col2", "col3");
""");
	}

	/// <inheritdoc />
	protected override void Down(MigrationBuilder migrationBuilder)
	{
		migrationBuilder.Sql("""
ALTER TABLE "table_name"
	DROP CONSTRAINT constraint_name;
""");
	}
}
```

**Comment by Snazzie** [comment]
+1

**Comment by borisdj** [comment]
Sure, will take a look in the following days.

**Comment by borisdj** [comment]
Fixed, new nuget will be pulished soon

**Comment by maximebono** [comment]
> Fixed, new nuget will be pulished soon

Hi @borisdj , do you have an estimation when the new nuget with this fix will be pulished ?

Thx ;)

**Comment by borisdj** [comment]
Published 8.1.1.

**Comment by Sekators** [comment]
> Fixed, new nuget will be pulished soon

Hi! 

You mentioned in the comments that the '**CountUniqueConstrain**' method is no longer used:
```
    /// <summary>
    /// Generates SQL query to count the unique constranints -  Not used, insted used only: CountUniqueIndex
    /// </summary>
    /// <param name="tableInfo"></param>
    public static string CountUniqueConstrain(TableInfo tableInfo)
```
However, this method is still used in 'CheckHasExplicitUniqueConstrainAsync':
```
internal static async Task<(bool, bool)> CheckHasExplicitUniqueConstrainAsync(DbContext context, TableInfo tableInfo,
        bool isAsync, CancellationToken cancellationToken)
    {
        string countUniqueConstrain = PostgreSqlQueryBuilder.CountUniqueConstrain(tableInfo);
```
Because of this, if my table already has a unique index but without a CONSTRAINT, an additional temporary unique index is still created, which duplicates the previous one.

Is this behavior correct?
